receiver/redisreceiver/documentation.md

```diff
@@ -35,8 +38,10 @@ These are the metrics available for this scraper.
 | **redis.rdb.changes_since_last_save** | Number of changes since the last dump |  | Sum(Int) | <ul> </ul> |
 | **redis.replication.backlog_first_byte_offset** | The master offset of the replication backlog buffer |  | Gauge(Int) | <ul> </ul> |
 | **redis.replication.offset** | The server's current replication offset |  | Gauge(Int) | <ul> </ul> |
+| **redis.role** | Redis node's role, 0 means replica, 1 means primary |  | Gauge(Int) | <ul> <li>role</li> </ul> |
```

Comment:
This is basically using a Gauge as an Enum. I don't think at this time OpenTelemetry does endorse the use of Gauges as Enum values, but suggest to use a Counter with a string attribute `role`. This way, if your backend supports aggregating Sums (which most backends do, I think), you should be able to answer questions like "How many nodes are running in Role `primary`?". There are multiple reasons why sending an enum as a gauge might not be the best choice, https://github.com/open-telemetry/opentelemetry-specification/issues/1711 gives a good overview.

Reply:
I see. In this case, this might be an UpDownCounter (which would be represented as a non-monotonic sum). Lets say we export Cumulative Temporality, although that is a technicality that should not impact how it works. I picture it like this: 
Every node that runs the redis server emits the metric with the following attributes:

| Interval  | Attributes               | Value |
| --------- | ------------------------ | ----- |
| T_0 - T_1 | `node=1`, `role=primary` | 1     |
| T_0 - T_1 | `node=1`, `role=replica` | 0     |
|           |                          |       |
| T_1 - T_2 | `node=1`, `role=primary` | 1     |
| T_1 - T_2 | `node=1`, `role=replica` | 0     |
|           |                          |       |
| T_2 - T_3 | `node=1`, `role=primary` | 0     |
| T_2 - T_3 | `node=1`, `role=replica` | 1     |

T_0 - T_1, T_1 - T_2, and T_2 - T_3 are three export intervals. The metric is exported three times in the example above. All of the exports share the same metric name, so I omitted it in the table. Each export contains two data points, one per respective value of `role`.

In the first two intervals, the node emits `1` for the `role=primary` and `0` for `role=replica`. In the third interval it emits `0` for `role=primary` and `1` for `role=replica`. Paired with the `node` dimension, you can track the state changes for this particular node. If don't have the `node` attribute, you get an overall picture of all nodes that are currently running, e.g. "there are three nodes running as `primary` and 5 nodes running as `replica`" without you having to deduce anything. 

Deducing would also require knowing what `0` and `1` _means_, which might be fine on the Collector (since we can hardcode it there), but not all backends allow for interpreting Gauges as state or setting up the respective mapping. 

If you want to model state changes ("The node switched roles, and I want to be notified about it immediately"), that seems more like a case for an Event (which is similar to a Log, and will actually be transported using the OTLP log data model). At this time, the Events API is not finished (https://github.com/open-telemetry/oteps/pull/202).

There is also a (somewhat lengthy) discussion about the Enum as Gauge topic here: https://github.com/open-telemetry/opentelemetry-specification/pull/2518#discussion_r863131948

Reply:
According to ` In OTLP, a Sum point can represent a monotonic count or a non-monotonic count. This means an OTLP Sum is either translated into a Timeseries Counter, when the sum is monotonic, or a Gauge when the sum is not monotonic.`
If it's a non-monotic sum type, it will be exported as gauge metric, so why can't we just define it as gauge at the beginning?

Reply:
Yeah, it's a bit confusing. *UpDownCounter* is to be used for things that evolve by observing changes. If Redis was perfectly observable, it would report when the role of a node changes from *primary* to *replica* and vice versa. For each change, it would push new values for the metric:
* node is no longer primary: `redis.role{state="primary"} -1`
* node is now replica: `redis.role{state="replica"} 1`

Same goes for a filesystem perfectly observable: it would report a increase in usage when a file is written to it, by the size of that file.

However, the instrumentation of "something" is rarely perfectly observable, and we can only get the "current" absolute value. That's why most receivers use an *Asynchronous UpDownCounter* which allows to set the absolute value, just like a *Gauge*.

It's almost philosophical.

Reply:
When it's exported as a gauge metric, what's the value? For example, if the value is 0 when it's replica and 1 when it's primary. Will it be the sum of all data points at that interval? If it emitted 3 times as primary in that interval, will the gauge value be 3?

Reply:
> According to `In OTLP, a Sum point can represent a monotonic count or a non-monotonic count. This means an OTLP Sum is either translated into a Timeseries Counter, when the sum is monotonic, or a Gauge when the sum is not monotonic.`

This quote is taken from the [Consumer recommendation](https://opentelemetry.io/docs/reference/specification/metrics/datamodel/#opentelemetry-protocol-data-model-consumer-recommendations) section. It seems to be aimed at transformations that need to be made in case the consumer does not know how to deal with the OTLP data as-is. I am sure there will be backends that will accept non-monotonic sums, and treat them differently from Gauges.

The difference, as far as I understand it, is that Gauge uses `LastValue` semantics by default, while monotonic Sum uses `Sum` semantics. Let's consider the following example: 
You have 5 redis nodes, two of them primary and 3 replicas. That means you would get metrics like these (sent from each of the nodes individually):


| `node` | `role`    | value | timestamp |
| ------ | --------- | ----- | --------- |
| `1`    | `primary` | `1`   | `2`       |
| `2`    | `replica` | `1`   | `4`       |
| `3`    | `replica` | `1`   | `6`       |
| `4`    | `primary` | `1`   | `1`       |
| `5`    | `replica` | `1`   | `8`       |

I left out the `0`s for readability (e. g. there would also be a data point that has `node=1` and `role=replica` with a value of `0`, but I don't think it's important for this example).

Let's say you are only interested in how many nodes are running in which role, and you want this information for 10-second timeframes (the nodes probably don't all report at the exact same time). In this case, you would remove the `node` attribute, and use data that was recorded between timestamps 0 and 10.

If you use Gauge, and last value semantics, the last recorded value for each attribute set will be used, so there would be:

| `role`    | `value` |  
| --------- | ------- | 
| `primary` | `1`     | 
| `replica` | `1`     |

Which would not be very interesting. It just gives you the information that "the last node to report its data in the timeframe reported that it was `role=primary` by setting the value to `1`". Note that the outcome is dependent only on _when_ the value arrived on the collector, and nothing else. If you use the Gauge in the way you proposed it (which is different from what I am proposing here) it also depends on which node reported last. The gauge would tell you "the last reported role in this timeframe was `primary`", but would drop all other data. Both of these do not make much sense IMHO. 

 If you use Sums, you would get 

| `role`    | `value` |
| --------- | ------- |
| `primary` | `2`     |
| `replica` | `3`     |

because it would sum together all values for the timeframe (and also add the 0s that I omitted earlier). If something changes (remove one primary and add a replica) the counts would be updated for the next interval.

In both cases (use as Gauge and use as Sum), the `value` **never** carries any information about the role of the nodes, this is always _only_ done by attributes. The value only counts how many times a certain set of Attributes is present.

To come back to your initial comment: If your backend does not support non-monotonic sums, you could export the values from the last table above as gauges (and then they would make sense), but it is crucial that the data is in Sum form while it is transformed here on the collector.

Reply:
I agree with @pirgeo . This should be `UpDownCounter` with the role set as an attribute, not encoded as a gauge value.

Reply:
Sure I will make that change